[sbt 1.0] Bumping up to Zinc 1.0.0-M1

eed3si9n opened the pull request:
Bumping up to new modules.
- Some changes to the DSL organization according to http://www.scala-sbt.org/0.13/docs/Coding-Guideline.html
- Minor adjustment to class discovery code - 738e8fb2f395f79364cb450e97c50d0db0331921

/review @Duhemm, @dwijnand

Duhemm committed:
Update to util 0.1.0-M11

Duhemm committed:
Upgrade to latest zinc

eed3si9n committed:
Bumping up dependencies

eed3si9n committed:
Adjust discovery code to new Zinc

Duhemm commented on MIGRATION.md:
````diff
@@ -3,3 +3,9 @@ Migration notes
+- change import in auto plugin to:
+
+```scala
+import sbt._, syntax._, Keys._
````
Should we advertise the existence of `sbt.Import`? That's what we use to keep the source compatibility of build definitions, but I guess we don't want to keep it around.

eed3si9n replied:
`Import` is now a trait, and I'm mixing that into the package object.

gkossakowski commented:
Zinc dropped support for Scala 2.8 and 2.9 which causes this test failure:

```
[info] [info] 'compiler-bridge_2.10' not yet compiled for Scala 2.8.1.final. Compiling...
[error] /tmp/sbt_ce7f5e4/xsbt/API.scala:43: error: ';' expected but string literal found.
[error]           s"$className:\n\t${names.mkString(", ")}"
[error]            ^
[error] /tmp/sbt_ce7f5e4/xsbt/Analyzer.scala:33: error: ')' expected but string literal found.
[error]               assert(sym.isClass, s"${sym.fullName} is not a class")
```

gkossakowski commented on main/actions/src/main/scala/sbt/DotGraph.scala:
```diff
@@ -30,7 +30,7 @@ object DotGraph {
     def file(name: String) = new File(outputDir, name)
     IO.createDirectory(outputDir)
-    generateGraph(file("int-source-deps"), "dependencies", relations.internalSrcDep, sourceToString, sourceToString)
+    generateGraph(file("int-source-deps"), "dependencies", relations.internalClassDep, identity[String], identity[String])
```
This should be renamed to `int-class-deps`.

dwijnand commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -508,18 +509,18 @@ object Defaults extends BuildCommon {
         val ans: Seq[Analysis] = cp.flatMap(_.metadata get Keys.analysis) map { case a0: Analysis => a0 }
         val succeeded = TestStatus.read(succeededFile(s.cacheDirectory))
-        val stamps = collection.mutable.Map.empty[File, Long]
+        val stamps = collection.mutable.Map.empty[String, Long]
```
Is this change expected? `File` isn't the most impressive of types, but it still beats `String` in my book.

gkossakowski replied:
Stamps were tracked per source file but now are tracked per class (that is named by a String). Tracking stamps by class name is more natural when everything else is tracked by class name (as it is in zinc).

dwijnand replied:
I see, thanks. But that's not intuitive. I'd advise to consider a ClassName type.

dwijnand commented:
Not sure how I feel about having to import the second half of things with `syntax`, but LGTM.

eed3si9n committed:
Fix apiinfo/show-circular-structure

eed3si9n committed:
Remove int-ant-style test

eed3si9n committed:
Remove run2.8 test

eed3si9n committed:
Fix compiler-project/semantic-errors

eed3si9n commented:
There seems to be legitimate problem around nested auto plugin discovery.. I get `java.lang.ClassNotFoundException: example.Imports.B$` when the plugin looks like

``` scala
package example

object Imports {
  object A extends AutoPlugin
}
```

while the classes directory contains `/private/var/folders/7g/l96y5q310h90xsz109qqjpmc0000gn/T/sbt_dc68f035/auto-plugins/project/target/scala-2.11/sbt-1.0.0-SNAPSHOT/classes/example/Imports$B$.class`

Ref sbt/zinc#116?

eed3si9n committed:
Fix project/auto-import

eed3si9n committed:
Fix project/binary-plugin

eed3si9n committed:
Fix project/flatten

eed3si9n committed:
Fix project/global-plugin

eed3si9n committed:
Fix project/isolated-build-definitions

eed3si9n committed:
Fix source-dependencies/cross-source

eed3si9n committed:
Disable source-dependencies/import-class

eed3si9n committed:
Old incremental compiler is no more

eed3si9n committed:
Remove trait-private-var test from sbt/sbt

eed3si9n committed:
Remove trait-super test from sbt/sbt

eed3si9n committed:
Fixes source-dependencies/transitive-memberRef

eed3si9n committed:
Fix test framework tests

eed3si9n committed:
Fix test quick

eed3si9n committed:
Fixes project/auto-plugins and add pending test for nested

There seems to be an issue with the discovery of auto plugins in an
inner nested class. The error you would get is:

[info] java.lang.ClassNotFoundException: sbttest.Imports.A$
[info] 	at java.net.URLClassLoader$1.run(URLClassLoader.java:202)
[info] 	at java.security.AccessController.doPrivileged(Native Method)
[info] 	at java.net.URLClassLoader.findClass(URLClassLoader.java:190)
[info] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:306)
[info] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:247)
[info] 	at java.lang.Class.forName0(Native Method)
[info] 	at java.lang.Class.forName(Class.java:249)
[info] 	at
sbt.internal.inc.ModuleUtilities$.getObject(ModuleUtilities.scala:16)
[info] 	at
sbt.internal.inc.ModuleUtilities$.getCheckedObject(ModuleUtilities.scala
:22)
[info] 	at
sbt.internal.inc.ModuleUtilities$$anonfun$getCheckedObjects$1.apply(Modu
leUtilities.scala:25)
[info] 	at
sbt.internal.inc.ModuleUtilities$$anonfun$getCheckedObjects$1.apply(Modu
leUtilities.scala:25)
[info] 	at scala.collection.immutable.Stream.map(Stream.scala:418)
[info] 	at
sbt.internal.inc.ModuleUtilities$.getCheckedObjects(ModuleUtilities.scal
a:25)
[info] 	at sbt.PluginDiscovery$.loadModules(PluginDiscovery.scala:138)
[info] 	at
sbt.PluginDiscovery$.binarySourceModules(PluginDiscovery.scala:132)
[info] 	at sbt.PluginDiscovery$.discover$1(PluginDiscovery.scala:36)
[info] 	at sbt.PluginDiscovery$.discoverAll(PluginDiscovery.scala:45)
[info] 	at sbt.internal.Load$.loadPlugins(Load.scala:856)

eed3si9n committed:
Renamed per review

gkossakowski commented on main/actions/src/main/scala/sbt/Tests.scala:
```diff
@@ -273,7 +273,19 @@ object Tests {
     discover(frameworks flatMap TestFramework.getFingerprints, allDefs(analysis), log)
 
-  def allDefs(analysis: CompileAnalysis) = analysis match { case analysis: Analysis => analysis.apis.internal.values.flatMap(_.api.definitions).toSeq }
+  def allDefs(analysis: CompileAnalysis) = analysis match {
```
This method used to return top level classes and objects but now returns also nested and inherited. Is that change intentional?

Have considered using the following definition?

``` scala
def allDefs(analysis: CompileAnalysis): Seq[xsbti.api.ClassLike] = {
  def companionsApis(c: xsbti.api.Companions) = Seq(c.classApi, c.objectApi)
  analysis match {
    case analysis: Analysis =>
      analysis.apis.internal.values.flatMap(x => companionsApis(x.api)).toSeq

  }
}
```

eed3si9n replied:
sbt 0.13 is able to detect auto plugins nested in another object.

gkossakowski replied:
Everywhere in `PluginDiscovery` different methods explicitly talk about top-level modules. Do you have a sense how plugins declared in another object are discovered?
